pkg/imgpkg/registry/keychain.go

```diff
@@ -0,0 +1,214 @@
+// Copyright 2020 VMware, Inc.
+// SPDX-License-Identifier: Apache-2.0
+package registry
+
+import (
+	"fmt"
+	"strings"
+	"sync"
+	"time"
+
+	regauthn "github.com/google/go-containerregistry/pkg/authn"
+	regname "github.com/google/go-containerregistry/pkg/name"
+)
+
+type KeychainOpts struct {
+	Username string
+	Password string
+	Token    string
+	Anon     bool
+}
+
+func Keychain(keychainOpts KeychainOpts, environFunc func() []string) regauthn.Keychain {
+	return regauthn.NewMultiKeychain(customRegistryKeychain{opts: keychainOpts},
```

Comment:
it seems that customRegistryKeychain falls back to retrying default keychain. this would happen before trying to access envKeychain based on my understanding of NewMultiKeychain. are the tests falling to envKeychain because regauthn.DefaultKeychain is set to anon? this seems like a pretty nasty global variable.

Reply:
Also if you set Anon to true, it will try the next keychain in the list

Reply:
How does the following sound:

## case 1) username + password is provided via cli flags
MultiKeychain uses customKeyChain

## case 2) anon is provided via cli flags
An "Anonymous" keychain is used (notice that MultiKeychain isn't used)

## case 3) env credentials are provided with the anon cli flag
e.g.
```
REGISTRY_HOSTNAME=some.reg
REGISTRY_USERNAME=foo
imgpkg pull --anon
```
An "Anonymous" keychain is used (notice that MultiKeychain isn't used)

## case 4) env credentials are provided, *and* credentials also provided via cli flags, *and* credentials are also contained in `config.json`
MultiKeychain traverses the keychain (custom keychain -> env keychain -> default keychain) until it finds a keychain to resolve the resource

Reply:
> An "Anonymous" keychain is used (notice that MultiKeychain isn't used)

im not sure we would want that. e.g. you might be configuring anonymous access for everything except for providing credentials for a specific registry. it seems at high level that MultiKeyChain does not capture "i found anon case and i want to use it" and instead treats anon case as fallthrough to the next option. how can we continue using MultiKeyChain and support our "explicit" anon case?

Reply:
Ok.

I'll make the ordering the following:

```
1. registry matching feature
  a. IMGPKG_USERNAME_0
2. generic auth
  a. IMGPKG_USERNAME
  b. --username <-- would take precedence over (a)
3. anon flag
4. default
```